lib/rucio/core/replica.py

```diff
@@ -704,6 +706,41 @@ def _resolve_dids(dids, unavailable, ignore_availability, all_states, resolve_ar
     return file_clause, dataset_clause, state_clause, files, constituents
 
 
+def _pick_n_random(nrandom, generator):
+    """
+    Select n random elements from the generator
+    """
+
+    if not nrandom:
+        # pass-through the data unchanged
+        yield from generator
+        return
+
+    # A "reservoir sampling" algorithm:
+    # Copy the N first files from the generator. After that, following element may be picked to substitute
+    # one of the previously selected element with a probability which decreases as the number of encountered elements grows.
```

Comment:
I like it. Intuitively, when looking at it first, it seems that the first results have a much higher probability of selection, but in the end, as long as you touch all elements, it's uniform. 👍